DIPs/DIP1NNN-DK.md

```diff
@@ -0,0 +1,688 @@
+This section contains five examples that demonstrate use cases for the proposed changes.
+Each example *wrongly uses `@trusted`* under *current semantics* followed by a `main` function demonstrating why.
+After this DIP is implemented, these will all become correct usages of `@trusted` and the `main` functions will fail to compile.
+
+### A tagged union
+This example shows how to safely overlap an unsafe type with another type.
+```D
+struct Var {
+    private @system union {
+        string asString = "var";
+        double asDouble;
+    }
+    enum Type : byte {
+        Str,
+        Num,
+    }
+    private @system Type type;
+
+    this(double x) @trusted {
+        this.type = Type.Num;
+        this.asDouble = x;
+    }
+
+    this(string s) @trusted {
+        this.type = Type.Str;
+        this.asString = s;
+    }
+
+    double getDouble() const @trusted {
+        return type == Type.Str ? asDouble : double.nan;
+    }
+
+    string getString() const @trusted {
+        return type == Type.Str ? asString : null;
+    }
+}
+
+import std;
+
+void main() @safe {
+    Var v = 1.0;
+    writeln(v.asString); // memory corruption: the string's length (and pointer
+    // on 32-bit targets) have been meddled with
+}
+```
+
+### Using uninitialized memory
+This simple stack data structure shows how to safely use uninitialized memory.
+In this case the stack data structure is using uninitialized stack memory, but the same idea applies to using malloc'ed memory.
+```D
+private struct Stack(T, int size = 32) {
+    private @system T[size] stack = void;
+    private @system int _top = 0;
+
+    T pop() @trusted {if (empty) assert(0); return stack[--_top];}
+    void push(T x) @trusted {if (full) assert(0); stack[_top++] = x;}
+    bool empty() const {return _top == 0;}
+    bool full() const {return _top == size;}
+}
+
+auto getStack(T)() @trusted {
+    Stack!T result = void;
+    result._top = 0;
+    return result;
+}
+
+import std;
+
+void main() @safe {
+    auto stack = getStack!string();
+    stack._top += 1; // not allowed after this DIP
+    writeln(stack.pop); // memory corruption: garbage string is printed
+}
+```
+
+### A custom allocator
+Here we make a simple 'bump the pointer' allocator.
+```D
+@system private ubyte[4096] heap;
+@system private int heapIndex;
+
+/// allocate an array of T
+/// never frees the memory
+T[] customAlloc(T)(int length) {
+    // round up heapIndex to multiple of T.alignof
+    heapIndex = (heapIndex + T.alignof-1) & ~(T.alignof-1);
+    auto result = cast(T[]) heap[heapIndex..heapIndex + T.sizeof*length];
+    heapIndex += heapIndex + T.sizeof*length;
+    return result;
+}
+
+import std;
+
+void main() @safe {
+    string[] strArr = customAlloc!string(3);
+    heapIndex = 0; // mess up allocator integrity! not allowed after this DIP.
+    int[] intArr = customAlloc!int(3);
+    intArr[2] = 0x8035FDF0; // overwrites string array
+    writeln(strArr[0]); // memory corruption: constructed pointer
+}
+```
```

Comment:
The text above says that "each example wrongly uses `@trusted`", but there's no `@trusted` in this example.

Reply:
Forgot to test this one, thanks.